### ISSUE TITLE
Ubuntu 12.04 Python 2.7.3 pythonnet Installation Issue

### ISSUE THREAD

**Issue by laskeym** [pythonnet/pythonnet#508]
### Environment

-   Pythonnet version: 2.3.0
-   Python version: 2.7.3
-   Operating System: Ubuntu 12.04

### Details

-   Installing pythonnet on Ubuntu 12.04.
-   Followed [https://travis-ci.org/pythonnet/pythonnet/jobs/148513880](url) build.

### Error

`sudo python setup.py build_ext --inplace
running build_ext
Checking for updates from https://www.nuget.org/api/v2/.
Currently running NuGet.exe 4.1.0.
NuGet.exe is up to date.
MSBuild auto-detection: using msbuild version '14.0' from '/usr/lib/mono/xbuild/14.0/bin'.
All packages listed in packages.config are already installed.

>>>> xbuild tool is deprecated and will be removed in future updates, use msbuild instead <<<<

XBuild Engine Version 14.0
Mono, Version 5.0.1.1
Copyright (C) 2005-2013 Various Mono authors

>>>> xbuild tool is deprecated and will be removed in future updates, use msbuild instead <<<<

XBuild Engine Version 14.0
Mono, Version 5.0.1.1
Copyright (C) 2005-2013 Various Mono authors
                Configuration: ReleaseMono Platform: x64
runtime.cs(284,17): warning CS0162: Unreachable code detected
Configuration: ReleaseMono Platform: x64
Configuration: ReleaseMono Platform: x64
dynamic.cs(121,28): warning CS0618: 'PythonEngine.RunString(string, IntPtr?, IntPtr?)' is obsolete: 'RunString is deprecated and will be removed. Use Exec/Eval/RunSimpleString instead.'
TestPythonEngineProperties.cs(171,17): warning CS0162: Unreachable code detected
Configuration: ReleaseMono Platform: x64`

Not sure as to what is going on except that there is some unreachable code.

**Comment by denfromufa** [comment]
Where do you see the error? I only see warnings.

**Comment by laskeym** [comment]
I'd say yes seeing as my `pip freeze` command does not output pythonnet.  I saw your reply in the other thread(sorry for the double post), I'm going to take a look at that wiki and see if it helps any.

**Comment by laskeym** [comment]
I also tried running it this way:

`pip install pythonnet -U`

Collecting pythonnet
  Using cached pythonnet-2.3.0.tar.gz
Building wheels for collected packages: pythonnet
  Running setup.py bdist_wheel for pythonnet ... error
  Complete output from command /home/dclark/.virtualenvs/BPM/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-E2ihW1/pythonnet/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/tmpzuIBDmpip-wheel- --python-tag cp27:
  running bdist_wheel
  running build
  running build_ext
  Unable to obtain lock file access on '/tmp/NuGetScratch/lock/e2b5d5bdf8ef11e5965b095073313fdd02112550' for operations on '/home/dclark/.config/NuGet/NuGet.Config'. This may mean that a different user or administator is holding this lock and that this process does not have permission to access it. If no other process is currently performing an operation on this file it may mean that an earlier NuGet process crashed and left an inaccessible lock file, in this case removing the file '/tmp/NuGetScratch/lock/e2b5d5bdf8ef11e5965b095073313fdd02112550' will allow NuGet to continue.
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/pip-build-E2ihW1/pythonnet/setup.py", line 405, in <module>
      zip_safe=False,
    File "/usr/lib/python2.7/distutils/core.py", line 152, in setup
      dist.run_commands()
    File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
      self.run_command(cmd)
    File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
      cmd_obj.run()
    File "/home/dclark/.virtualenvs/BPM/local/lib/python2.7/site-packages/wheel/bdist_wheel.py", line 179, in run
      self.run_command('build')
    File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
      self.distribution.run_command(command)
    File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
      cmd_obj.run()
    File "/usr/lib/python2.7/distutils/command/build.py", line 128, in run
      self.run_command(cmd_name)
    File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
      self.distribution.run_command(command)
    File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
      cmd_obj.run()
    File "/usr/lib/python2.7/distutils/command/build_ext.py", line 339, in run
      self.build_extensions()
    File "/usr/lib/python2.7/distutils/command/build_ext.py", line 448, in build_extensions
      self.build_extension(ext)
    File "/tmp/pip-build-E2ihW1/pythonnet/setup.py", line 139, in build_extension
      self._install_packages()
    File "/tmp/pip-build-E2ihW1/pythonnet/setup.py", line 269, in _install_packages
      subprocess.check_call(cmd, shell=use_shell)
    File "/usr/lib/python2.7/subprocess.py", line 511, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command 'mono tools/nuget/nuget.exe update -self' returned non-zero exit status 1

  ----------------------------------------
  Failed building wheel for pythonnet
  Running setup.py clean for pythonnet
Failed to build pythonnet
Installing collected packages: pythonnet
  Running setup.py install for pythonnet ... error
    Complete output from command /home/dclark/.virtualenvs/BPM/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-E2ihW1/pythonnet/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-LX7jM7-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/dclark/.virtualenvs/BPM/include/site/python2.7/pythonnet:
    running install
    running build
    running build_ext
    Unable to obtain lock file access on '/tmp/NuGetScratch/lock/e2b5d5bdf8ef11e5965b095073313fdd02112550' for operations on '/home/dclark/.config/NuGet/NuGet.Config'. This may mean that a different user or administator is holding this lock and that this process does not have permission to access it. If no other process is currently performing an operation on this file it may mean that an earlier NuGet process crashed and left an inaccessible lock file, in this case removing the file '/tmp/NuGetScratch/lock/e2b5d5bdf8ef11e5965b095073313fdd02112550' will allow NuGet to continue.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-E2ihW1/pythonnet/setup.py", line 405, in <module>
        zip_safe=False,
      File "/usr/lib/python2.7/distutils/core.py", line 152, in setup
        dist.run_commands()
      File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/home/dclark/.virtualenvs/BPM/local/lib/python2.7/site-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/usr/lib/python2.7/distutils/command/install.py", line 601, in run
        self.run_command('build')
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/distutils/command/build.py", line 128, in run
        self.run_command(cmd_name)
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/distutils/command/build_ext.py", line 339, in run
        self.build_extensions()
      File "/usr/lib/python2.7/distutils/command/build_ext.py", line 448, in build_extensions
        self.build_extension(ext)
      File "/tmp/pip-build-E2ihW1/pythonnet/setup.py", line 139, in build_extension
        self._install_packages()
      File "/tmp/pip-build-E2ihW1/pythonnet/setup.py", line 269, in _install_packages
        subprocess.check_call(cmd, shell=use_shell)
      File "/usr/lib/python2.7/subprocess.py", line 511, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command 'mono tools/nuget/nuget.exe update -self' returned non-zero exit status 1

    ----------------------------------------
Command "/home/dclark/.virtualenvs/BPM/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-E2ihW1/pythonnet/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-LX7jM7-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/dclark/.virtualenvs/BPM/include/site/python2.7/pythonnet" failed with error code 1 in /tmp/pip-build-E2ihW1/pythonnet/

**Comment by denfromufa** [comment]
Have you tried this:

`sudo pip install git+https://github.com/pythonnet/pythonnet -U --force`

?

**Comment by laskeym** [comment]
`sudo pip install git+https://github.com/pythonnet/pythonnet -U --force`
Usage: /usr/bin/pip install [OPTIONS] PACKAGE_NAMES...

/usr/bin/pip install: error: no such option: --force

So I tried it without sudo:
`pip install git+https://github.com/pythonnet/pythonnet -U --force`

Collecting git+https://github.com/pythonnet/pythonnet
  Cloning https://github.com/pythonnet/pythonnet to /tmp/pip-hQdXe_-build
Installing collected packages: pythonnet
  Running setup.py install for pythonnet ... error
    Complete output from command /home/dclark/.virtualenvs/BPM/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-hQdXe_-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-cDywyZ-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/dclark/.virtualenvs/BPM/include/site/python2.7/pythonnet:
    running install
    running build
    running build_ext
    Unable to obtain lock file access on '/tmp/NuGetScratch/lock/e2b5d5bdf8ef11e5965b095073313fdd02112550' for operations on '/home/dclark/.config/NuGet/NuGet.Config'. This may mean that a different user or administator is holding this lock and that this process does not have permission to access it. If no other process is currently performing an operation on this file it may mean that an earlier NuGet process crashed and left an inaccessible lock file, in this case removing the file '/tmp/NuGetScratch/lock/e2b5d5bdf8ef11e5965b095073313fdd02112550' will allow NuGet to continue.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-hQdXe_-build/setup.py", line 408, in <module>
        zip_safe=False,
      File "/usr/lib/python2.7/distutils/core.py", line 152, in setup
        dist.run_commands()
      File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/home/dclark/.virtualenvs/BPM/local/lib/python2.7/site-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/usr/lib/python2.7/distutils/command/install.py", line 601, in run
        self.run_command('build')
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/distutils/command/build.py", line 128, in run
        self.run_command(cmd_name)
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/distutils/command/build_ext.py", line 339, in run
        self.build_extensions()
      File "/usr/lib/python2.7/distutils/command/build_ext.py", line 448, in build_extensions
        self.build_extension(ext)
      File "/tmp/pip-hQdXe_-build/setup.py", line 142, in build_extension
        self._install_packages()
      File "/tmp/pip-hQdXe_-build/setup.py", line 272, in _install_packages
        subprocess.check_call(cmd, shell=use_shell)
      File "/usr/lib/python2.7/subprocess.py", line 511, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command 'mono tools/nuget/nuget.exe update -self' returned non-zero exit status 1

    ----------------------------------------
Command "/home/dclark/.virtualenvs/BPM/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-hQdXe_-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-cDywyZ-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/dclark/.virtualenvs/BPM/include/site/python2.7/pythonnet" failed with error code 1 in /tmp/pip-hQdXe_-build/

**Comment by denfromufa** [comment]
ok, what do `pip --version` and `sudo pip --version` show? Can you update both (if there are 2) since you probably have ancient pip? `pip install -U pip` and `sudo pip install -U pip`. Also you may be able to install without sudo, but need to use `--user` option to pip, otherwise you don't have the permissions for installation.

instead of `--force` you can also try `--force-reinstall`.

**Comment by laskeym** [comment]
Didn't know sudo had it's own pip version.  When running `sudo pip install git+https://github.com/pythonnet/pythonnet -U --force` it successfully installs pythonnet and I do see it in `sudo pip freeze`.  So it sounds like when I run my python script, I should be using sudo.

I also tried `pip install git+https://github.com/pythonnet/pythonnet -U --force --user` and the output was:
Can not perform a '--user' install. User site-packages are not visible in this virtualenv.

I tried `pip install git+https://github.com/pythonnet/pythonnet -U --force-reinstall` and received the error in my previous comment.
